### PR TITLE
#3800 Improve inventory import 

### PR DESCRIPTION
**Pull request by cristinamghita**
#3800 Improve inventory import

### CODE REVIEW

File: de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5490540_gh3800_ExtractLocatorDimensions.sql [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	v_locatorDimensions.locatorValue := textInput;
+	v_locatorDimensions.locatorX := v_dimensions[2];
+	v_locatorDimensions.locatorY := v_dimensions[3];
+	v_locatorDimensions.locatorZ := v_dimensions[4];
+	v_locatorDimensions.locatorX1 := v_dimensions[5];
+	
+	return v_locatorDimensions;
+END;
+$BODY$
+LANGUAGE plpgsql VOLATILE
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5490540_gh3800_ExtractLocatorDimensions.sql]: why volatile?

File: de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+				.append("SET M_Locator_ID=(SELECT MAX(M_Locator_ID) FROM M_Locator l ")
+				.append("WHERE i.LocatorValue=l.Value AND i.AD_Client_ID=l.AD_Client_ID) ")
+				.append("WHERE M_Locator_ID IS NULL AND LocatorValue IS NOT NULL ")
+				.append("AND I_IsImported<>'Y' ")
+				.append(whereClause);
+		int no = DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.debug("Set Locator from Value =" + no);
+	}
+
+	private void dbCreateLocators()
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java]: what if on 2 different I_Inventory records you have same locator (i.e. same x,y,z,x1 etc) ?

File: de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	{
+		final I_M_Locator locator = InterfaceWrapperHelper.newInstance(I_M_Locator.class);
+		locator.setAD_Org_ID(importRecord.getAD_Org_ID());
+		locator.setM_Warehouse_ID(importRecord.getM_Warehouse_ID());
+		locator.setValue(importRecord.getLocatorValue());
+		locator.setX(importRecord.getX());
+		locator.setY(importRecord.getY());
+		locator.setZ(importRecord.getZ());
+		locator.setX1(importRecord.getX1());
+		locator.setDateLastInventory(importRecord.getDateLastInventory());
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java]: what's DateLastInventory and why setting it only when the locator is created?

File: de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+		logger.debug("Set Product from Value=" + no);
+
+		sql = new StringBuilder("UPDATE I_Inventory i ")
+				.append("SET M_Product_ID=(SELECT MAX(M_Product_ID) FROM M_Product p ")
+				.append(" WHERE i.UPC=p.UPC AND i.AD_Client_ID=p.AD_Client_ID) ")
+				.append("WHERE M_Product_ID IS NULL AND UPC IS NOT NULL ")
+				.append("AND I_IsImported<>'Y' ")
+				.append(whereClause);
+		no = DB.executeUpdate(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.debug("Set Product from UPC=" + no);
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java]: pls don't see string concatenation when logging. pls check in other places too.

File: de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java [hunk 5]
```diff
@@ -275,6 +278,43 @@ public I_M_AttributeInstance getCreateAttributeInstance(final I_M_AttributeSetIn
 		return instanceNew;
 	}
 
+	@Override
+	public void setAttributeInstanceValue(@NonNull final I_M_AttributeSetInstance asi, @NonNull final I_M_Attribute attribute, @NonNull final Object value)
+	{
+		final I_M_AttributeInstance attributeInstance = Services.get(IAttributeDAO.class).retrieveAttributeInstance(asi, attribute.getM_Attribute_ID());
+		if (attributeInstance == null)
+		{
+			throw new AdempiereException("instanceExisting not exists");
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java]: if attribute instance was not found, why not creating it?

File: de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java [hunk 6]
```diff
@@ -275,6 +278,43 @@ public I_M_AttributeInstance getCreateAttributeInstance(final I_M_AttributeSetIn
...
+		if (X_M_Attribute.ATTRIBUTEVALUETYPE_Date.equals(attributeValueType))
+		{
+			attributeInstance.setValueDate(Env.parseTimestamp(value.toString()));
+		}
+		else if (X_M_Attribute.ATTRIBUTEVALUETYPE_List.equals(attributeValueType))
+		{
+			attributeInstance.setValue(value.toString());
+
+			final I_M_AttributeValue attributeValue = Services.get(IAttributeDAO.class).retrieveAttributeValueOrNull(attribute, value.toString());
+			attributeInstance.setM_AttributeValue(attributeValue);
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/AttributeSetInstanceBL.java]: i think setting the M_AttributeValue_ID is not needed. see de.metas.modelvalidator.M_AttributeInstance.updateAttributeValueIfList(I_M_AttributeInstance)

File: de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java (hunk 2 above)
Comment (cristinamghita) [reply de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java]: is fine, same locator, probably different product

File: de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5490540_gh3800_ExtractLocatorDimensions.sql (hunk 1 above)
Comment (cristinamghita) [reply de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5490540_gh3800_ExtractLocatorDimensions.sql]: by mistake

File: de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java (hunk 3 above)
Comment (cristinamghita) [reply de.metas.business/src/main/java/org/adempiere/impexp/inventory/MInventoryImportTableSqlUpdater.java]: hm.. good point; I need to update also

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/interceptor/M_Inventory.java [hunk 7]
```diff
@@ -86,9 +98,55 @@ else if (qtyDiff.signum() > 0)
...
 		}
 	}
 
+	private final void transferAttributesToHU(
+			@NonNull final I_M_InventoryLine inventoryLine,
+			@NonNull final I_M_HU hu)
+	{
+		final IHUTrxBL huTrxBL = Services.get(IHUTrxBL.class);
+		final IContextAware contextInitial = InterfaceWrapperHelper.getContextAware(inventoryLine);
+		final IHUContextProcessorExecutor executor = huTrxBL.createHUContextProcessorExecutor(contextInitial);
```
Comment (teosarca) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/interceptor/M_Inventory.java]: i think u can safely use de.metas.handlingunits.hutransaction.IHUTrxBL.createHUContextProcessorExecutor()

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/interceptor/M_Inventory.java [hunk 8]
```diff
@@ -86,9 +98,55 @@ else if (qtyDiff.signum() > 0)
...
+				return IHUContextProcessor.NULL_RESULT; // can't transfer from nothing
+			}
+			final IAttributeStorage huAttributeStorageTo = attributeStorageFactory.getAttributeStorage(hu);
+
+			final IHUStorageFactory storageFactory = huContext.getHUStorageFactory();
+			final IHUStorage huStorageFrom = storageFactory.getStorage(hu);
+
+			final IHUAttributeTransferRequest request = new HUAttributeTransferRequestBuilder(huContext)
+					.setProduct(inventoryLine.getM_Product())
+					.setQty(inventoryLine.getQtyInternalUse())
```
Comment (teosarca) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/interceptor/M_Inventory.java]: i think u shall use inventoryBL.getMovementQty(inventoryLine)

**Pull request merged**